open-apparel-registry/open-apparel-registry
# Identify exact matches pre-dedupe

## Overview

When matching facilities, look for exact matches prior to entering
the dedupe process. Exact matches (for name, address, and country)
can be automatically matched, and don't need to enter the dedupe
process.

Adds new fields to FacilityListItems to store the cleaned name and cleaned address, allowing efficient filtering for exact matches. 

If multiple exact matches exist, we prefer first the active list items; then items submitted by the same contributor; then finally the most recent items. 

Connects #1549 

## Demo

<img width="461" alt="Screen Shot 2022-01-05 at 9 27 44 AM" src="https://user-images.githubusercontent.com/21046714/148239768-8fb2af43-f7e2-420c-99bc-d666e82ab4de.png">

## Notes

PPE handling was not implemented for exact-match items. 

## Testing Instructions

* Run migrations. 
* Login as c2@example.com, and navigate to http://localhost:8081/api/docs/#!/facilities/facilities_create. Submit an exact copy of an existing facility with `create=true`:
```
{
"name": "Alif Embroidery Village Ltd.",
"address": "Bangobandhu Road, Tangabari, Ashulia, Savar, Dhaka 1341",
"country": "Bangladesh"
}
```
* Confirm that the item has matched. Note the facility id. In an incognito browser, login as c1@example.com and navigate to the new FacilityListItem in the [Django Admin](http://localhost:8081/admin/api/facilitylistitem). In the processing results, you should see `exact_match: true`.
* As c1@example.com, navigate to [adjust facility matches](http://localhost:6543/dashboard/adjustfacilitymatches). Transfer the match you just created from the facility it matched to to any other facility. 
* Resubmit the same facility via the Swagger API. It should now match to the facility you transferred the FacilityListItem to. 
* Deactivate both newly created matches. Submit the same facility details via CSV.
[duplicates_test.csv](https://github.com/open-apparel-registry/open-apparel-registry/files/7815785/duplicates_test.csv)
* Process the list:
```
./scripts/manage batch_process --list-id 16 --action parse
./scripts/manage batch_process --list-id 16 --action geocode
./scripts/manage batch_process --list-id 16 --action match
```
* Confirm that the newly uploaded item matched to the original facility via exact match. 

## Checklist
